### PR TITLE
Feature/apm 121

### PR DESCRIPTION
**Pull request by cpjulia**
### Scope & Purpose

Added support for the UPSERT query to have indexHint as an extra parameter for 
OPTIONS. It will be used as a hint by the inner FOR loop that is performed
as part of the UPSERT query, and would help in cases such as UPSERT not 
picking the best index automatically for lookup. Example:

```js
UPSERT { a: 1234 } 
  INSERT { a: 1234, name: "Abcd"} 
  UPDATE {name: ""} IN myCollection 
  OPTIONS {indexHint: "hash_a"};
```

Docs PR: https://github.com/arangodb/docs/pull/776
Jira ticket: https://arangodb.atlassian.net/browse/APM-121

- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [x] Docs PR: https://github.com/arangodb/docs/pull/776
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket number: https://arangodb.atlassian.net/browse/APM-121
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** in shell_server_aql
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: arangod/Aql/grammar.h [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/* A Bison parser, made by GNU Bison 3.7.6.  */
```
Comment (dothebart) [inline arangod/Aql/grammar.h:1]: hm, generated with an older bison?

File: tests/js/server/aql/aql-index-hints.js [hunk 2]
```diff
@@ -421,7 +421,38 @@ function ahuacatlSkiplistOverlappingTestSuite () {
       assertEqual(usedIndexes[0][0], alternateSortingIndex);
     },
 
+    testUpsertWithIndexHint: function () {
+      const prefix = " UPSERT { a: 1234 } INSERT { a: 1234 } UPDATE {} IN " + cn + " OPTIONS ";
+      const indexHints = [
+        [{}, ["hash_a", "hash_a_b", "skip_a", "skip_a_b"]],
```
Comment (dothebart) [inline tests/js/server/aql/aql-index-hints.js:427]: hm, without MMFiles, these are merely names - @jsteemann any plan when we will drop these aliases?

File: arangod/Aql/grammar.h (hunk 1 above)
Comment (cpjulia) [reply arangod/Aql/grammar.h:1]: Strange that I have recently installed it in my computer, I'll try to update it and generate the files again and push here

Comment (cpjulia) [reply arangod/Aql/grammar.h]: Recompiled it

File: tests/js/server/aql/aql-index-hints.js (hunk 2 above)
Comment (cpjulia) [reply tests/js/server/aql/aql-index-hints.js:427]: these names were used for preserving the behavior of the rest of the tests, all the tests in this file use the same collection and indexes, so I used the same ones for leaving it more uniform. 
It doesn't compromise the functionality of this specific test, which was to see if the indexes are being used actually as index hint

**Review by mchacki — APPROVED** [review]
LGTM

**Review by mchacki — APPROVED** [review]
LGTM

**Pull request merged**